### PR TITLE
WIP: [CARBONDATA-1839] [DataLoad]Fix bugs in compressing sort temp files

### PR DESCRIPTION
**Pull request by xuchuanyin**
Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [X] Any interfaces changed?
  `YES, ONLY CHANGE INTERNAL INTERFACES`
 - [X] Any backward compatibility impacted?
  `NO`
 - [X] Document update required?
  `YES`
 - [X] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        `ADDED TESTS`
        - How it is tested? Please attach test report.
        `TESTED IN LOCAL CLUSTER`
        - Is it a performance related change? Please attach the performance test report.
        `YES`
        - Any additional information to help reviewers in testing this change.
        `There are some duplicate code in write temp sort files found during this bug fixing and I plan to optimize it in successive PR not in this one.`
 - [X] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
        `NOT RELATED`

RESOLVE
===
1. fix bugs in compressing sort temp files

2. optimize parameters

Use an object `TableFieldStat` to wrap about 6 parameters. This object can be used to further optimize the code and I will use it in other PRs.

3. add tests

Add tests to enable sort_temp_file_compressed while doing data loading

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/401/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2061/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1674/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/429/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2085/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1700/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2095/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/445/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1714/

File: processing/src/main/java/org/apache/carbondata/processing/loading/sort/impl/ParallelReadMergeSorterImpl.java [hunk 1]
```diff
@@ -80,11 +81,7 @@ public void initialize(SortParameters sortParameters) {
         File.separator, CarbonCommonConstants.SORT_TEMP_FILE_LOCATION);
     finalMerger =
         new SingleThreadFinalSortFilesMerger(dataFolderLocations, sortParameters.getTableName(),
-            sortParameters.getDimColCount(),
-            sortParameters.getComplexDimColCount(), sortParameters.getMeasureColCount(),
-            sortParameters.getNoDictionaryCount(), sortParameters.getMeasureDataType(),
-            sortParameters.getNoDictionaryDimnesionColumn(),
-            sortParameters.getNoDictionarySortColumn());
+            new TableFieldStat(sortParameters));
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/loading/sort/impl/ParallelReadMergeSorterImpl.java]: I think the indentation is not correct

File: processing/src/main/java/org/apache/carbondata/processing/loading/sort/unsafe/holder/UnsafeSortTempFileChunkHolder.java [hunk 2]
```diff
@@ -352,6 +330,49 @@ private void fillDataForPrefetch() {
     }
   }
 
+  private Object[][] getBatchedRowFromStream(int expected) throws CarbonSortKeyAndGroupByException {
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/loading/sort/unsafe/holder/UnsafeSortTempFileChunkHolder.java]: please add comment for this function

File: processing/src/main/java/org/apache/carbondata/processing/sort/sortdata/SortTempFileChunkHolder.java [hunk 3]
```diff
@@ -381,6 +351,51 @@ private void fillDataForPrefetch() {
     return holder;
   }
 
+  private Object[][] getBatchedRowFromStream(int expected) throws CarbonSortKeyAndGroupByException {
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/sort/sortdata/SortTempFileChunkHolder.java]: please add comment for this function

File: processing/src/main/java/org/apache/carbondata/processing/sort/sortdata/TableFieldStat.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.carbondata.processing.sort.sortdata;
+
+import java.util.Objects;
+
+import org.apache.carbondata.core.metadata.datatype.DataType;
+
+public class TableFieldStat {
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/sort/sortdata/TableFieldStat.java]: Add comment

**Comment by xuchuanyin** [comment]
@jackylk Thanks for your review, since another PR #1594 has been merged, I'll optimize this PR to reduce duplicate codes in writing sort temp files.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/495/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1757/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2141/

**Comment by jackylk** [comment]
Is the rework completed for this PR?

**Comment by xuchuanyin** [comment]
No, it is a little complex to reduce the duplicate code. I'll make a full test as possible as I can.

**Comment by xuchuanyin** [comment]
@jackylk I close this PR and raise #1632 instead.

**Pull request closed without merging**